DependencyInjection/Compiler/DependencyBuilderPass.php

```diff
@@ -0,0 +1,208 @@
+use OAT\Library\Lti1p3Core\Registration\RegistrationRepositoryInterface;
+use OAT\Library\Lti1p3Core\Security\Key\KeyChain;
+use OAT\Library\Lti1p3Core\Security\Key\KeyChainRepositoryInterface;
+use OAT\Library\Lti1p3Core\Tool\Tool;
+use Symfony\Component\Config\Definition\Processor;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+
+class DependencyBuilderPass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container)
+    {
+        $configuration = $this->processConfiguration($container);
+
+        // build & inject key chains definitions
+        $keyChainsDefinitions = $this->buildKeyChainsDefinitions($configuration);
+
+        $keyChainRepository = $container->getDefinition(KeyChainRepositoryInterface::class);
+        $keyChainRepository->setArgument(0, $keyChainsDefinitions);
+
+        // build & inject registrations definitions
+        $registrationsDefinitions = $this->buildRegistrationsDefinitions(
+            $configuration,
+            $keyChainsDefinitions,
+            $this->buildPlatformsDefinitions($configuration),
+            $this->buildToolsDefinitions($configuration)
+        );
+
+        $registrationRepository = $container->getDefinition(RegistrationRepositoryInterface::class);
+        $registrationRepository->setArgument(0, $registrationsDefinitions);
+    }
+
+    private function processConfiguration(ContainerBuilder $container): array
+    {
+        $processor = new Processor();
+
+        return $processor->processConfiguration(
+            new Configuration(),
+            $container->getExtensionConfig(Lti1p3Extension::ALIAS)
+        );
+    }
+
+    /**
+     * @return Definition[]
+     */
+    private function buildKeyChainsDefinitions(array $configuration): array
+    {
+        $definitions = [];
+
+        foreach ($configuration['key_chains'] as $keyId => $keyData) {
+            $definition = new Definition(KeyChain::class, [
+                $keyId,
+                $keyData['key_set_name'],
+                $keyData['public_key'],
+                $keyData['private_key'],
+                $keyData['private_key_passphrase']
+            ]);
+
+            $definitions[$keyId] = $definition;
+        }
+
+        return $definitions;
+    }
+
+    /**
+     * @return Definition[]
+     */
+    private function buildPlatformsDefinitions(array $configuration): array
+    {
+        $definitions = [];
+
+        foreach ($configuration['platforms'] as $platformId => $platformData) {
+            $definition = new Definition(Platform::class, [
+                $platformId,
+                $platformData['name'],
+                $platformData['audience'],
+                $platformData['oidc_authentication_url'] ?? null,
+                $platformData['oauth2_access_token_url'] ?? null
```

Comment:
not relevant anymore, since now it's delegated to builders.
Since those builders can be called if wanted also not from a compiler pass, I leave the ?? null to ease it's usage.